### ISSUE TITLE
resnet-50 vanilla performance not good enough

### ISSUE THREAD

**Issue by gngdb** [BayesWatch/tf-variational-dropout#6]
Trying to match the performance of the [resnet-50 reported in pytorch](https://github.com/kuangliu/pytorch-cifar). The learning rate, weight decay, batch size and [convolutional architecture](https://gist.github.com/gngdb/ca72d2c2b501d787fc677c54d94debe1) match. Unfortunately, the final accuracy does not. I'm only able to get 91% training in tensorflow, versus 93%+ in pytorch.

Other factors that might not match:

* Exact batch norm configuration
* Weight initialisation
* Training set augmentation

**Comment by gngdb** [comment]
Weight Initialisation
---------------------------

Both sample from uniform distributions and then scale.

The pytorch implementation [uses the default](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/conv.py#L40-L47). The scaling is `1./sqrt(c_in*k*k)`.

The tensorflow version uses `xavier_initializer_conv2d` by default, which [is reported in the docs to be](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/xavier_initializer): `sqrt(6. / (in + out))`. So, we'd assume that means: `sqrt(6. / (c_in*k*k + c_out*k*k))`. For exactness, [I checked the code](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/contrib/layers/python/layers/initializers.py#L140-L144): `math.sqrt(3.0 * factor / n)`, where `factor = ` and `n = (fan_in + fan_out) / 2.0`.

`fan_in` appears to be `c_in*k*k` (assuming tensorflow conv params are stored `(k,k,c_out,c_in)` and likewise  `fan_out = c_in*k*k`.

So the initialisations do differ quite a bit.

Fixed this in the most recent commit, adding `PytorchInit`.

**Comment by gngdb** [comment]
Batchnorm config
----------------

In tensorflow, we use [batch_norm with default params](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/batch_norm). Going to assume the placeholder `is_train` is working; assuming this isn't a problem with running averages being calculated on the test set. Pytorch is also using batch norm with [defaults](http://pytorch.org/docs/master/nn.html#torch.nn.BatchNorm2d).

Checking these defaults:

1. **tf**: `decay=0.999`, **pytorch**: `momentum=0.1`. Presumably `decay = 1-momentum`. So, default decay in tensorflow should be changed to 0.9 to match.
2. **tf**: `center=True`, **pytorch**: ???
3. **tf**: `scale=False`, **pytorch**: ???
4. **tf**: `epsilon=0.001`, **pytorch**: `eps=1e-5`. Not super important, but might as well make them match. For numerical stability.
5. **tf**: `batch_weights=True`, **pytorch**: ???
6. **tf**: `trainable=True`, **pytorch**: `affine=True`

Important note, code did not have the following in order to update the batch norm running mean. This seems very likely to have been something that could have caused problems:

```
  update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
  with tf.control_dependencies(update_ops):
      train_op = optimizer.minimize(loss)
```

**Comment by gngdb** [comment]
Training set augmentation
-------------------------------------

Tensorflow code in this repo does the following for `distrorted_inputs`:

* Randomly crop `32x32` out of `32x32` image. So, it's not actually cropping anything (my fault).
* `random_flip_left_right`
* `random_brightness`
* `random_contrast`
* `per_image_standardization`

[pytorch-cifar][pc] uses the following:

* Random crop with padding of 4 pixels.
* RandomHorizontalFlip
* Global Normalize with the following parameters: `(0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)`

Adding the padding and switching to global normalization.

[pc]: https://github.com/kuangliu/pytorch-cifar/blob/master/main.py

**Comment by gngdb** [comment]
Doing all of the above, performance plateaus at 92.5% accuracy on the test set.

**Comment by gngdb** [comment]
Seems likely a problem with regularisation, this tensorflow model is able to push the cross entropy loss much closer to zero in the final iterations than the pytorch model was ever able to do.

Therefore, next two things to check:

* Are we running *exactly* the same number of steps, globally?
* Further than just checking the convolutional layers, do the numbers of convolutional layers match? And are they exactly the same in the order they are called?
